README.md
# modstore-publisher
PHP CLI tool which allows to publish zipped MODX extra to modstore.pro marketplace

## TODO:
- [ ] Add symfony application and CLI component
- [ ] Add binary to easy run of tool
- [ ] Convert existing script to the proper class (or command)
- [ ] Make possible to configure it by params and options
- [ ] Write explicit documentation about usage in readme.md


changelog.md
# [0.3.0](https://github.com/alroniks/modstore-publisher/compare/v0.2.0...v0.3.0) (2021-05-14)


### Bug Fixes

* option descriptions moved to one place ([ce327c4](https://github.com/alroniks/modstore-publisher/commit/ce327c47f12170361242d5e5145e6e941440d7aa))


### Features

* improve formatting messages with SymfonyStyles ([d477c26](https://github.com/alroniks/modstore-publisher/commit/d477c269fd9d76c65f004bd9bb065ee5b59c7f8a))
* make possible configure package in interactive mode ([b10e1e7](https://github.com/alroniks/modstore-publisher/commit/b10e1e7cef12b4fa0542f3c4e4e04af8857d4e20))
* package path option instead of argument for better validation ([d1de0f4](https://github.com/alroniks/modstore-publisher/commit/d1de0f4a63ff334a022ecbb95075b4e97174b130))
* parsing extra page to get id and versions ([e65b723](https://github.com/alroniks/modstore-publisher/commit/e65b723e6bdd82c60715586399b154dffe084cac))
* require php 8+ and didom html parser ([e1e2622](https://github.com/alroniks/modstore-publisher/commit/e1e26225391aa56d4d5f854bfc04f6ebbac7f625))
* validation of mandatory options ([3f2e307](https://github.com/alroniks/modstore-publisher/commit/3f2e3078bc4a09d57b714dc860ed939fe02ec1f6))



# [0.2.0](https://github.com/alroniks/modstore-publisher/compare/v0.1.0...v0.2.0) (2021-04-27)


### Bug Fixes

* default value for modx and php required versions ([d21d502](https://github.com/alroniks/modstore-publisher/commit/d21d502ffe798059bea31412ad477632caaa52b2))


### Features

* configure command by options and params ([6cbee29](https://github.com/alroniks/modstore-publisher/commit/6cbee290742e030d85d83254e32badf44b84b92d))
* missed token exception ([e2ebba3](https://github.com/alroniks/modstore-publisher/commit/e2ebba3dc02fc090aaf7605d6b09e2737c90a080))
* parsing file signature to get name and version ([9a34771](https://github.com/alroniks/modstore-publisher/commit/9a34771a43cb625745a54caa6843cd3ef7682aa4))
* require html parser library ([b75b55d](https://github.com/alroniks/modstore-publisher/commit/b75b55dbac43b12703d48345300d4faf81ed0890))



# 0.1.0 (2021-04-27)





src/Command/PublishCommand.php
<?php
/**
 * This file is part of modstore publisher package.
 * For complete copyright and license information, see the LICENSE file
 * found in the top-level directory of this distribution.
 **/

declare(strict_types = 1);

namespace Alroniks\Publisher\Command;

use Alroniks\Publisher\ExtraException;
use Alroniks\Publisher\PackageForm;
use Alroniks\Publisher\SignatureException;
use Alroniks\Publisher\TokenException;
use DiDom\Document;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use JetBrains\PhpStorm\ArrayShape;
use JsonException;
use Psr\Http\Message\ResponseInterface;
use RuntimeException;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Console\Style\SymfonyStyle;

final class PublishCommand extends Command
{
    public const OPTION_PACKAGE = 'package';

    public const OPTION_LOGIN = 'login';
    public const OPTION_PASSWORD = 'password';

    public const OPTION_DEPRECATE = 'deprecate';
    public const OPTION_OVERRIDE = 'override';

    public const OPTION_REQUIRED_PHP_VERSION = 'php-version';
    public const OPTION_REQUIRED_MODX_VERSION = 'modx-version';
    public const OPTION_REQUIRED_MODX_VERSION_MAX = 'modx-version-until';

    public const OPTION_CHANGELOG = 'changelog';
    public const OPTION_CHANGELOG_ENGLISH = 'changelog-english';

    private const DESCRIPTIONS_MAP = [
        self::OPTION_PACKAGE => 'Path to the archive with compiled MODX package.',
        self::OPTION_LOGIN => 'User name (email) for login on modstore.',
        self::OPTION_PASSWORD => 'Password for login on modstore.',
        self::OPTION_CHANGELOG => 'Path to the file with changelog entries.',
        self::OPTION_CHANGELOG_ENGLISH => 'Alternative file for English variant in case, when original changelog on Russian. By default one file used for both cases.',
        self::OPTION_REQUIRED_MODX_VERSION => 'Minimal version of MODX which required for running the package code.',
        self::OPTION_REQUIRED_MODX_VERSION_MAX => "Up to what maximum version of MODX the package code is guaranteed to work.\n May be useful to limit packages, which are not compatible with MODX 3 yet.",
        self::OPTION_REQUIRED_PHP_VERSION => 'Minimal version of PHP which required for running the package code.',
        self::OPTION_DEPRECATE => "Disable all previous versions of the package.\n",
        self::OPTION_OVERRIDE => "Override existing version by the new binary package.\n If the version does not exist, the flag will be ignored.\n",
    ];

    private const MANDATORY_OPTIONS = [
        self::OPTION_PACKAGE,
        self::OPTION_LOGIN,
        self::OPTION_PASSWORD,
        self::OPTION_CHANGELOG
    ];

    private const SUPPORTS_VERSION_DEFAULT_MODX = '2.8';
    private const SUPPORTS_VERSION_DEFAULT_PHP = '7.2';

    // move to connector service
    private const CLIENT_BASE_URI = 'https://modstore.pro/';
    private const CLIENT_ENTRY_POINT = 'assets/components/extras/action.php';

    // todo: move to connector service
    private const ACTION_CREATE = 'office/versions/create';
    private const ACTION_UPDATE = 'office/versions/update';

    // minimum_supports
    // todo: parse lists with options from modstore, not hard defined
    private const MODX_VERSIONS = ['2.2', '2.3', '2.4', '2.5', '2.6', '2.7', '2.8', '3.0'];

    // supports
    // todo: parse lists with options from modstore, not hard defined
    private const MODX_VERSIONS_UNTIL = ['2.2', '2.3', '2.4', '2.5', '2.6', '2.7', '2.8', '3.0'];

    // todo: parse lists with options from modstore
    private const PHP_VERSIONS = ['5.3', '5.4', '5.5', '5.6', '7.0', '7.1', '7.2'];

    private Client $client;
    private SymfonyStyle $io;

    protected function configure(): void
    {
        $this
            ->setName('publish')
            ->setDescription('Publishes or updates the next package version to the marketplace.')
            ->setDefinition(
                [
                    // package
                    new InputOption(
                        self::OPTION_PACKAGE, 'e',
                        InputOption::VALUE_REQUIRED,
                        self::DESCRIPTIONS_MAP[self::OPTION_PACKAGE]
                    ),

                    // credentials
                    new InputOption(
                        self::OPTION_LOGIN, 'u',
                        InputOption::VALUE_REQUIRED,
                        self::DESCRIPTIONS_MAP[self::OPTION_LOGIN]
                    ),
                    new InputOption(
                        self::OPTION_PASSWORD, 'p',
                        InputOption::VALUE_REQUIRED,
                        self::DESCRIPTIONS_MAP[self::OPTION_PASSWORD]
                    ),

                    // changelog
                    new InputOption(
                        self::OPTION_CHANGELOG, null,
                        InputOption::VALUE_REQUIRED,
                        self::DESCRIPTIONS_MAP[self::OPTION_CHANGELOG]
                    ),
                    new InputOption(
                        self::OPTION_CHANGELOG_ENGLISH, null,
                        InputOption::VALUE_REQUIRED,
                        self::DESCRIPTIONS_MAP[self::OPTION_CHANGELOG_ENGLISH]
                    ),

                    // versions
                    new InputOption(
                        self::OPTION_REQUIRED_MODX_VERSION, null,
                        InputOption::VALUE_REQUIRED,
                        self::DESCRIPTIONS_MAP[self::OPTION_REQUIRED_MODX_VERSION],
                        self::SUPPORTS_VERSION_DEFAULT_MODX
                    ),
                    new InputOption(
                        self::OPTION_REQUIRED_MODX_VERSION_MAX, null,
                        InputOption::VALUE_REQUIRED,
                        self::DESCRIPTIONS_MAP[self::OPTION_REQUIRED_MODX_VERSION_MAX],
                    ),
                    new InputOption(
                        self::OPTION_REQUIRED_PHP_VERSION, null,
                        InputOption::VALUE_REQUIRED,
                        self::DESCRIPTIONS_MAP[self::OPTION_REQUIRED_PHP_VERSION],
                        self::SUPPORTS_VERSION_DEFAULT_PHP
                    ),

                    // flags
                    new InputOption(
                        self::OPTION_DEPRECATE, 'd', InputOption::VALUE_NONE,
                        self::DESCRIPTIONS_MAP[self::OPTION_DEPRECATE],
                    ),
                    new InputOption(
                        self::OPTION_OVERRIDE, 'r', InputOption::VALUE_NONE,
                        self::DESCRIPTIONS_MAP[self::OPTION_OVERRIDE],
                    ),
                ]
            )
            ->setHelp(
                 <<<EOT
The <info>publish</info> command connects to marketplace, analizes versions
and publishes or updates the version of the package in the repository.

It is default command, so no need to define it, just run the following command to get interactive mode.

<info>bin/publisher</info>

Read more in <href=https://github.com/alroniks/modstore-publisher#readme>Documentation</>

EOT
                );
    }

    public function initialize(InputInterface $input, OutputInterface $output): void
    {
        $this->io = new SymfonyStyle($input, $output);

        $this->client = new Client([
            'base_uri' => self::CLIENT_BASE_URI,
            'cookies' => true
        ]);

        parent::initialize($input, $output);
    }

    public function getIO(): SymfonyStyle
    {
        return $this->io;
    }

    public function getClient(): Client
    {
        return $this->client;
    }

    /**
     * @throws \Spatie\DataTransferObject\Exceptions\UnknownProperties
     * @throws \DiDom\Exceptions\InvalidSelectorException
     * @throws \JsonException
     */
    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $io = $this->getIO();

        $options = array_filter($input->getOptions(), static fn ($option) => !empty($option));
        $common = array_intersect( self::MANDATORY_OPTIONS, array_keys($options));

        if (count($common) !== count(self::MANDATORY_OPTIONS)) {
            $io->error('Required options are missed');
            $io->block('To continue, define following options with values, please.');

            $missed = array_chunk(
                array_filter(
                    self::DESCRIPTIONS_MAP,
                    static fn($key) => in_array($key, array_diff(self::MANDATORY_OPTIONS, $common), true),
                    ARRAY_FILTER_USE_KEY
                ),
                1,
                true
            );

            $io->definitionList('Missed options:', ...$missed);

            return self::FAILURE;
        }

        try {
            // 01. Attempt to login
            $this->formatServiceAnswer(
                $this->login(
                    $input->getOption(self::OPTION_LOGIN),
                    $input->getOption(self::OPTION_PASSWORD)
                )
            );

            // 02. Parse archive name
            [$package, $version] = $this->parsePackage($input->getOption(self::OPTION_PACKAGE));

            // 03. Get the page of the package
            $content = $this->getClient()
                ->request('GET', sprintf('office/packages/%s', $package))
                ->getBody()->getContents();

            // 04. Get the CSRF token
            $token = $this->parseToken($content);

            // 05. Get meta information about extra
            $extra = $this->parseExtra($content);

            //

            exit(0);

            // если версии совпадают и есть опция override, то нужно обновить
            // если опции нет - то просто добавиь
            // или спросить вопрос, перезаписать ли?

//            $form = new PackageForm(
//                action: self::ACTION_CREATE,
//                package_id: $extra['id']
//
//            );


//            print_r($extra);
//            print_r(array_column($extra['versions'], 'active'));

            // todo: confirm publishing on the last step
            // сводная таблица всех параметров - нужно брать из DTO

            // 07. Upload the compiled package
            $this->formatServiceAnswer(
                $this->upload($form, $token, true)
            );

            return self::SUCCESS;
        } catch (GuzzleException $gex) {
            $this->formatServiceAnswer($gex->getResponse());
            return self::FAILURE;
        } catch (SignatureException | TokenException | ExtraException $e) {
            $io->error($e->getMessage());
            return self::FAILURE;
        }
    }

    protected function interact(InputInterface $input, OutputInterface $output): void
    {
        $io = $this->getIO();

        $io->title('MODX Extra Publisher');
        $io->block('This tool will help you publish your package on modstore.pro marketplace.');
        $io->block('Follow the questions and to answer them to get package published.');

        if (!$input->getOption(self::OPTION_PACKAGE)) {
            $io->section('Defining path to the package');
            $io->text('Please, enter an valid path to file with compiled MODX package.');

            $input->setOption(self::OPTION_PACKAGE, $io->ask('Path to the file with package archive'));
        }

        if (!$input->getOption(self::OPTION_LOGIN) && !$input->getOption(self::OPTION_PASSWORD)) {

            $io->section('Signing in marketplace');
            $io->text('Please, enter login and password of your account on modstore.pro.');

            if (!$input->getOption(self::OPTION_LOGIN)) {
                $input->setOption(self::OPTION_LOGIN, $io->ask('Login'));
            }

            if (!$input->getOption(self::OPTION_PASSWORD)) {
                $input->setOption(self::OPTION_PASSWORD, $io->askHidden('Password'));
            }
        }

        $io->section('Configuring the uploading version');

        if (!$input->getOption(self::OPTION_CHANGELOG)) {
            $input->setOption(self::OPTION_CHANGELOG, $io->ask('Path to the file with changelog on Russian'));
        }

        if (!$input->getOption(self::OPTION_CHANGELOG_ENGLISH)
            && $io->confirm('Do you have different changelog for English language?', false)
        ) {
            $input->setOption(self::OPTION_CHANGELOG_ENGLISH, $io->ask('Path to file with changelog on English'));
        }

        if (!$input->getOption(self::OPTION_DEPRECATE)) {
            $input->setOption(self::OPTION_DEPRECATE, $io->confirm('Disable all previous versions of a package?', true));
        }
    }

    /**
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    private function login(string $email, string $password): ResponseInterface
    {
        return $this->getClient()->request(
            'POST',
            self::CLIENT_ENTRY_POINT,
            [
                'form_params' => [
                    'action' => 'office/login',
                    'email' => $email,
                    'password' => $password,
                ],
            ]
        );
    }

    private function parsePackage(string $path): array
    {
        $data = pathinfo($path);

        if ($data['extension'] !== 'zip') {
            throw new SignatureException('Provided file is not zip-archive.');
        }

        $pattern = '/^([a-z]+)-(\d+\.\d+\.\d+-[a-z]+)\.transport$/m';

        $check = preg_match($pattern, $data['filename'], $matches);

        if ($check === false) {
            throw new RuntimeException('Impossible to match package signature by regular expression.');
        } elseif ($check === 0) {
            throw new SignatureException('Provided file has an incompatible package name.');
        }

        return array_values(array_filter($matches, static fn($key) => in_array($key, [1, 2], true), ARRAY_FILTER_USE_KEY));
    }

    /**
     * @throws \DiDom\Exceptions\InvalidSelectorException
     */
    #[ArrayShape(['id' => "int", 'versions' => "array"])]
    private function parseExtra(string $content): array
    {
        $document = new Document($content);

        $input = $document->first('#office-package-form input[name="id"]');

        if (!$input) {
            throw new ExtraException('Element with extra ID is not found on the page.');
        }

        $versions = [];
        foreach ($document->find('#pdopage tr.version') as $version) {
            $cells = $version->find('td');

            $href = $cells[4]->first('a.fa-edit')->getAttribute('href');
            $segments = explode('/', $href);

            $versions[] = [
                'id' => (int) array_pop($segments),
                'active' => !str_contains($version->getAttribute('class'), 'inactive'),
                'version' => trim($cells[0]->text()),
                'released' => trim($cells[1]->text()),
                'downloads' => (int) trim($cells[2]->text()),
                'updated' => trim($cells[3]->text())
            ];
        }

        return [
            'id' => (int) $input->getAttribute('value'),
            'versions' => $versions
        ];
    }

    private function parseToken(string $content): string
    {
        // todo: replace by DiDom parser?
        $tempFile = tmpfile();
        fwrite($tempFile, $content);
        $metatags = get_meta_tags(stream_get_meta_data($tempFile)['uri']);

        if (!array_key_exists('csrf-token', $metatags)) {
            throw new TokenException('CSRF token not found, next operations impossible.');
        }

        return $metatags['csrf-token'];
    }

    /**
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    private function upload(PackageForm $form, string $token, bool $debug = false): ResponseInterface
    {
        return $this->getClient()->request('POST', self::CLIENT_ENTRY_POINT, [
            'multipart' => $form->multipart(),
            'headers' => ['X-CSRF-Token' => $token],
            'debug' => $debug
        ]);
    }

    /**
     * @throws JsonException
     */
    private function formatServiceAnswer(ResponseInterface $response): void
    {
        $io = $this->getIO();

        $decoded = json_decode($response->getBody()->getContents(), true, 512, JSON_THROW_ON_ERROR);
        $success = (bool) ($decoded['success'] ?? false);
        $message = (string) ($decoded['message'] ?? '');

        $success ?
            $io->success($message) :
            $io->error($message);
    }
}

